akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +37,156 @@ private object PoolSlot {
+          case x ⇒
+            log.error("invalid command {} when connected", x)
         }
       }
-      inflightRequests = immutable.Queue.empty
-      onNext(SlotEvent.Disconnected(slotIx, results.size) :: results)
-      if (canceled) onComplete()
 
-      context.become(unconnected)
-    }
+      // unconnected
+      override def onPush(): Unit = {
+        def createSubSourceOutlets() = {
+          connectionFlowSource = new SubSourceOutlet[HttpRequest]("RequestSource")
+          connectionFlowSource.setHandler(connectionOutFlowHandler)
 
-    override def onComplete(): Unit = {
-      exposedPublisher.shutdown(None)
-      super.onComplete()
-      shutdown()
-    }
+          connectionFlowSink = new SubSinkInlet[HttpResponse]("ResponseSink")
+          connectionFlowSink.setHandler(connectionInFlowHandler)
 
-    override def onError(cause: Throwable): Unit = {
-      exposedPublisher.shutdown(Some(cause))
-      super.onError(cause)
-      shutdown()
-    }
+          setHandler(in, connected)
+          isConnected = true
+        }
 
-    def shutdown(): Unit = context.stop(self)
-  }
+        grab(in) match {
+          case ConnectEagerlyCommand ⇒
+            createSubSourceOutlets()
 
-  private case class FromConnection(ev: Any) extends NoSerializationVerificationNeeded
-
-  private class FlowInportActor(slotProcessor: ActorRef) extends ActorPublisher[HttpRequest] with ActorLogging {
-    def receive: Receive = {
-      case ev: Request            ⇒ slotProcessor ! FromConnection(ev)
-      case OnNext(r: HttpRequest) ⇒ onNext(r)
-      case OnComplete             ⇒ onCompleteThenStop()
-      case OnError(e)             ⇒ onErrorThenStop(e)
-      case Cancel ⇒
-        slotProcessor ! FromConnection(Cancel)
-        context.stop(self)
-    }
-  }
-  def flowInportProps(s: ActorRef) = Props(new FlowInportActor(s)).withDeploy(Deploy.local)
-
-  private class FlowOutportActor(slotProcessor: ActorRef) extends ActorSubscriber with ActorLogging {
-    def requestStrategy = ZeroRequestStrategy
-    def receive: Receive = {
-      case Request(n) ⇒ request(n)
-      case Cancel     ⇒ cancel()
-      case ev: OnNext ⇒ slotProcessor ! FromConnection(ev)
-      case ev @ (OnComplete | OnError(_)) ⇒
-        slotProcessor ! FromConnection(ev)
-        context.stop(self)
+            Source.fromGraph(connectionFlowSource.source)
+              .via(connectionFlow).runWith(Sink.fromGraph(connectionFlowSink.sink))(subFusingMaterializer)
+
+            connectionFlowSink.pull()
+            pull(in)
+
+            push(out1, ConnectedEagerly(slotIx))
+
+          case DispatchCommand(rc: RequestContext) ⇒
+            createSubSourceOutlets()
+
+            inflightRequests.add(rc)
+
+            Source.single(rc.request).concat(Source.fromGraph(connectionFlowSource.source))
+              .via(connectionFlow).runWith(Sink.fromGraph(connectionFlowSink.sink))(subFusingMaterializer)
+
+            connectionFlowSink.pull()
```

Comment:
But why dispatch more requests onto a known-to-get-closed connection although we know they'll take a retry cycle anyway?